From 728cb2b2aed7058cfcf7eeee31af6a16dbeb791f Mon Sep 17 00:00:00 2001
From: fukiame <fukiame@proton.me>
Date: Tue, 29 Aug 2023 21:30:52 +0700
Subject: [PATCH 2/5] fs/open: add ksu faccessat hook

---
 fs/open.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/fs/open.c b/fs/open.c
index b14aef04ee01..ba3a3eb48841 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -348,6 +348,10 @@ SYSCALL_DEFINE4(fallocate, int, fd, int, mode, loff_t, offset, loff_t, len)
 	return ksys_fallocate(fd, mode, offset, len);
 }
 
+
+extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user, int *mode,
+			   int *flags);
+
 /*
  * access() needs to use the real uid/gid, not the effective uid/gid.
  * We do this by temporarily clearing all FS-related capabilities and
@@ -363,6 +367,8 @@ long do_faccessat(int dfd, const char __user *filename, int mode)
 	int res;
 	unsigned int lookup_flags = LOOKUP_FOLLOW;
 
+	ksu_handle_faccessat(&dfd, &filename, &mode, NULL);
+
 	if (mode & ~S_IRWXO)	/* where's F_OK, X_OK, W_OK, R_OK? */
 		return -EINVAL;
 
-- 
2.42.0

